### PR TITLE
[MXNET-684] Add `cond` operator

### PR DESCRIPTION
**Pull request by junrushao1994**
Waiting for the `while_loop` operator to be merged so that I could rebase to master. Please do not merge for now.

## Description ##

This PR is part of the proposal of adding a set of control flow operators to MXNet. [Link](https://cwiki.apache.org/confluence/display/MXNET/Optimize+dynamic+neural+network+models+with+control+flow+operators) to proposal. See also `foreach` (#11531) and `while_loop` (#11566).

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add `ifelse` operator in src/operator/control_flow.cc

### TODO ###
- [x] Wait for `while_loop` to be merged
- [x] Rebase to master
- [x] Fix flaky tests in `while_loop` because of numeric overflow
- [x] Remove duplicated benchmark codes

### CODE REVIEW

**Comment by junrushao1994** [comment]
@zheng-da @piiswrong @szha @eric-haibin-lin Hey could you help review this PR?

File: src/operator/control_flow.cc [hunk 1]
```diff
@@ -977,6 +913,342 @@ WhileLoopGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& og
...
+  CHECK_EQ(in_shape->size() + 3U, (size_t) params.num_args);
+  CHECK_EQ(out_shape->size(), (size_t) params.num_outputs);
+  CHECK_EQ(attrs.subgraphs.size(), 3U);
+  CHECK_EQ(attrs.subgraphs[0]->outputs.size(), 1U);
+  CHECK_EQ(attrs.subgraphs[1]->outputs.size(), attrs.subgraphs[2]->outputs.size());
+  // infer shape for cond, then and else
+  auto infer_subg = [&params, in_shape, out_shape](std::shared_ptr<Symbol> subg,
+                                                   ShapeVector *_subg_out,
+                                                   const nnvm::Tuple<dim_t> &input_locs,
+                                                   bool fill_out_shape) {
```
Comment (zheng-da) [inline src/operator/control_flow.cc]: can you also reuse this function?

File: src/operator/control_flow.cc [hunk 2]
```diff
@@ -977,6 +913,342 @@ WhileLoopGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& og
...
+  ShapeVector cond_out_shape{TShape(1U)};  // this means: [(1, )]
+  ShapeVector then_out_shape(params.num_outputs);
+  ShapeVector else_out_shape(params.num_outputs);
+  bool succ_0 = infer_subg(attrs.subgraphs[0], &cond_out_shape, \
+                           params.cond_input_locs, false);
+  bool succ_1 = infer_subg(attrs.subgraphs[1], &then_out_shape, \
+                           params.then_input_locs, true);
+  bool succ_2 = infer_subg(attrs.subgraphs[2], &else_out_shape, \
+                           params.else_input_locs, true);
+  return succ_0 && succ_1 && succ_2;
```
Comment (zheng-da) [inline src/operator/control_flow.cc]: you need to check `then_out_shape` and `else_out_shape` and see if they are the same.

File: src/operator/control_flow.cc [hunk 3]
```diff
@@ -977,6 +913,342 @@ WhileLoopGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& og
...
+  extract_by_loc(*in_type, params.cond_input_locs, &cond_in_type);
+  extract_by_loc(*in_type, params.then_input_locs, &then_in_type);
+  extract_by_loc(*in_type, params.else_input_locs, &else_in_type);
+  std::vector<int> cond_out_type = {0};
+  bool succ_0 = InferSubgraphDataType(*attrs.subgraphs[0], &cond_in_type, &cond_out_type);
+  CHECK(sync_in_in(params.cond_input_locs, in_type, &cond_in_type, is_udf));
+  bool succ_1 = InferSubgraphDataType(*attrs.subgraphs[1], &then_in_type, out_type);
+  CHECK(sync_in_in(params.then_input_locs, in_type, &then_in_type, is_udf));
+  bool succ_2 = InferSubgraphDataType(*attrs.subgraphs[2], &else_in_type, out_type);
+  CHECK(sync_in_in(params.else_input_locs, in_type, &else_in_type, is_udf));
```
Comment (zheng-da) [inline src/operator/control_flow.cc]: the same here.

File: src/operator/control_flow.cc [hunk 4]
```diff
@@ -977,6 +913,342 @@ WhileLoopGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& og
...
+  *dispatch_mode = DispatchMode::kFComputeEx;
+  bool succ_0 = InferSubgraphStorage(*attrs.subgraphs[0], dev_mask, \
+                                     &cond_mode, &cond_in_attrs, &cond_out_attrs);
+  CHECK(sync_in_in(params.cond_input_locs, in_attrs, &cond_in_attrs, is_udf));
+  bool succ_1 = InferSubgraphStorage(*attrs.subgraphs[1], dev_mask, \
+                                     &then_mode, &then_in_attrs, out_attrs);
+  CHECK(sync_in_in(params.then_input_locs, in_attrs, &then_in_attrs, is_udf));
+  bool succ_2 = InferSubgraphStorage(*attrs.subgraphs[2], dev_mask, \
+                                     &else_mode, &else_in_attrs, out_attrs);
+  CHECK(sync_in_in(params.else_input_locs, in_attrs, &else_in_attrs, is_udf));
```
Comment (zheng-da) [inline src/operator/control_flow.cc]: the same here.

**Comment by zheng-da** [comment]
do you need to rebase to the master since the flaky test has been fixed.

File: python/mxnet/ndarray/contrib.py [hunk 5]
```diff
@@ -363,3 +362,97 @@ def _func_wrapper(loop_vars):
                 ["  Step %d, shape is %s" % (i, str(x.shape)) for i, x in enumerate(items)]
             ))
     return stacked_outputs, list(loop_vars)
+
+def ifelse(cond, then_func, else_func, inputs):
+    """Run a if-then-else using user-defined condition and computation
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/mxnet/ndarray/contrib.py]: Fixed

File: python/mxnet/ndarray/contrib.py [hunk 6]
```diff
@@ -363,3 +362,97 @@ def _func_wrapper(loop_vars):
...
+
+    This operator simulates a if-like branch which chooses to do one of
+    the two customized computations according to the specified condition.
+
+    `inputs` is a list of NDArrays on which the condition and computations reply on.
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/mxnet/ndarray/contrib.py]: Fixed

File: python/mxnet/symbol/contrib.py [hunk 7]
```diff
@@ -556,3 +556,154 @@ def _union_inputs(*graphs):
     outputs = [result[i] for i in range(num_out_data)]
     final_loop_vars = [result[i] for i in range(num_out_data, num_outputs)]
     return outputs, final_loop_vars
+
+def ifelse(cond, then_func, else_func, inputs, name="ifelse"):
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/mxnet/symbol/contrib.py]: Fixed

File: python/mxnet/symbol/contrib.py [hunk 8]
```diff
@@ -556,3 +556,154 @@ def _union_inputs(*graphs):
...
+        if isinstance(inputs, Symbol):
+            inputs = (inputs, )
+        if not isinstance(inputs, tuple):
+            raise ValueError("%s must be a Symbol, or a tuple or list of Symbol" % (name, ))
+        for item in inputs:
+            if not isinstance(item, Symbol):
+                raise ValueError("%s must be a Symbol, or a tuple or list of Symbol" % (name, ))
+        return inputs
+
+    def _create_subgraph(graph_vars, graph_func, subgraph_name):
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/mxnet/symbol/contrib.py]: They are not exactly the same. One would search for var_locs, another doesn't.

File: python/mxnet/ndarray/contrib.py [hunk 9]
```diff
@@ -363,3 +362,97 @@ def _func_wrapper(loop_vars):
...
+    inputs = _to_ndarray_tuple(inputs, "inputs")
+    if len(inputs) == 0:
+        raise ValueError("inputs should contain at least one element")
+    branch = _to_python_scalar(cond(*inputs), bool, "Return value of cond")
+    if branch:
+        outputs = then_func(*inputs)
+        outputs = _to_ndarray_tuple(outputs, "outputs of then_func")
+    else:
+        outputs = else_func(*inputs)
+        outputs = _to_ndarray_tuple(outputs, "outputs of else_func")
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply python/mxnet/ndarray/contrib.py]: Let's give up ><

File: src/operator/control_flow.cc (hunk 1 above)
Comment (junrushao1994) [reply src/operator/control_flow.cc]: They are not identical either.

File: src/operator/control_flow.cc (hunk 2 above)
Comment (junrushao1994) [reply src/operator/control_flow.cc]: My bad, fixed

File: src/operator/control_flow.cc (hunk 3 above)
Comment (junrushao1994) [reply src/operator/control_flow.cc]: Two subgraphs write to the same `out_type`, so we don't have to worry in this case.

File: src/operator/control_flow.cc (hunk 4 above)
Comment (junrushao1994) [reply src/operator/control_flow.cc]: Two subgraphs write to the same `out_type`, so we don't have to worry in this case.

**Comment by junrushao1994** [comment]
We are going to rename the operator from `ifelse` to `cond` according to discussion with @zheng-da offline

File: python/mxnet/ndarray/contrib.py [hunk 10]
```diff
@@ -363,3 +362,97 @@ def _func_wrapper(loop_vars):
                 ["  Step %d, shape is %s" % (i, str(x.shape)) for i, x in enumerate(items)]
             ))
     return stacked_outputs, list(loop_vars)
+
+def ifelse(cond, then_func, else_func, inputs):
```
Comment (eric-haibin-lin) [inline python/mxnet/ndarray/contrib.py]: Why do we have both cond and inputs? Can we just have cond which could be true/false?

File: python/mxnet/ndarray/contrib.py [hunk 11]
```diff
@@ -363,3 +362,97 @@ def _func_wrapper(loop_vars):
                 ["  Step %d, shape is %s" % (i, str(x.shape)) for i, x in enumerate(items)]
             ))
     return stacked_outputs, list(loop_vars)
+
+def cond(cond_func, then_func, else_func, inputs):  # pylint: disable=redefined-outer-name
```
Comment (eric-haibin-lin) [inline python/mxnet/ndarray/contrib.py]: can we remove inputs and have tf style interface?

File: python/mxnet/ndarray/contrib.py (hunk 10 above)
Comment (zheng-da) [reply python/mxnet/ndarray/contrib.py]: to make it consistent with TF?
https://www.tensorflow.org/api_docs/python/tf/cond

File: python/mxnet/ndarray/contrib.py (hunk 11 above)
Comment (junrushao1994) [reply python/mxnet/ndarray/contrib.py]: Good idea. Just finished the API change. Thanks!

**Comment by junrushao1994** [comment]
@zheng-da and I decided to do an API change, according to the valuable comments from @eric-haibin-lin.

Here is the signature of our new API: `condition(cond: Symbol, then_func: Function, else_func: Function) -> List[Symbol]`, where `then_func` and `else_func` are functions with no arguments. 

Although there are no actual difference between the old and new APIs in the backend, we believe that this change will make customers easier to use our API.

Thanks again for @eric-haibin-lin and @zheng-da for the valuable discussion!

**Comment by junrushao1994** [comment]
So could someone help merge the code?

File: python/mxnet/ndarray/contrib.py (hunk 10 above)
Comment (junrushao1994) [reply python/mxnet/ndarray/contrib.py]: It may help improve user experience I think.

**Comment by junrushao1994** [comment]
We are trying to get this in 1.3. So could someone help merge this PR?

File: python/mxnet/ndarray/contrib.py [hunk 12]
```diff
@@ -28,7 +28,7 @@
 except ImportError:
     pass
 
-__all__ = ["rand_zipfian", "foreach", "while_loop"]
+__all__ = ["rand_zipfian", "foreach", "while_loop", "condition"]
```
Comment (piiswrong) [inline python/mxnet/ndarray/contrib.py]: `condition` is not a good name. Maybe `cond`, or `conditional` or `if_else` or something like that

File: python/mxnet/ndarray/contrib.py [hunk 13]
```diff
@@ -363,3 +362,87 @@ def _func_wrapper(loop_vars):
                 ["  Step %d, shape is %s" % (i, str(x.shape)) for i, x in enumerate(items)]
             ))
     return stacked_outputs, list(loop_vars)
+
+def condition(cond, then_func, else_func):
```
Comment (piiswrong) [inline python/mxnet/ndarray/contrib.py]: change signature to be the same with other packages?

**Comment by junrushao1994** [comment]
@piiswrong Our initial name is ifelse. @zheng-da proposed to change a new name ‘cond’. ‘cond’ cannot make pylint happy, because it conflicts with the first argument of ‘while_loop’. So then I change it to ‘condition’. So could you guys help let know me a best naming so that we could get it merged today? @zheng-da @piiswrong

**Comment by junrushao1994** [comment]
I change `condition` to `cond` for now

**Comment by junrushao1994** [comment]
Thank you so much guys for helping me with this PR, and offering me valuable suggestions!

**Pull request merged**